pulser/waveforms.py

```diff
@@ -73,25 +85,29 @@ def duration(self) -> int:
         """The duration of the pulse (in ns)."""
         pass
 
-    @property
+    @cached_property
     @abstractmethod
+    def _samples(self) -> np.ndarray:
+        pass
+
+    @property
     def samples(self) -> np.ndarray:
         """The value at each time step that describes the waveform.
 
         Returns:
             np.ndarray: A numpy array with a value for each time step.
         """
-        pass
+        return self._samples.copy()
```

Comment:
Cast this to `np.ndarray`.

Reply:
> Cast this to `np.ndarray`.

`pulser\waveforms.py:100: error: Redundant cast to "ndarray"`

It is not necessary since the return type is already included in the function def.

Reply:
But without it you have this error: 
```
pulser/waveforms.py:100: error: Returning Any from function declared to return "ndarray"
```

How come?

Reply:
The reason is because it is importing the `cached_property` with stubs, so `cached_property` is returning `Any`. My suggestion above might fix this.